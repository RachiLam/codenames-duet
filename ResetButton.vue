<template>
  <button @click="reset" />
</template>

<script>
export default {
  methods: {
    reset() {
      window.location.reload();
    },
  },
};
</script>

<style lang="scss" scoped>
  @import './variables';

  button {
    width: $reset-button-size;
    height: $reset-button-size;
    background-color: $accent-light;
    border-radius: 25%;
  }
</style>
