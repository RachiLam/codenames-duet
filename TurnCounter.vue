<template>
  <div
    class="turn-counter counter-like"
    :class="type"
    @click="$emit('click')"
  />
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      required: true,
    },
  },
};
</script>

<style lang="scss" scoped>
  @import './variables';

  .turn-counter {
    &.player {
      background-color: $primary;
    }

    &.remaining {
      background-color: $accent-dark;
    }

    &.error {
      background-color: $error;
    }
  }
</style>
