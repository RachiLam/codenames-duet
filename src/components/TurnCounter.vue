<template>
  <div
    class="turn-counter counter-like"
    :class="type"
    @click="$emit('click')"
  />
</template>

<script>
export default {
  props: {
    type: {
      type: String,
      required: true,
      validator: (value) => ['error', 'normal', 'player'].includes(value),
    },
  },
};
</script>

<style lang="scss" scoped>
  @use 'styles/colors';

  .turn-counter {
    &.error {
      background-color: colors.$error;
    }

    &.normal {
      background-color: colors.$accent-dark;
    }

    &.player {
      background-color: colors.$success;
    }
  }
</style>
