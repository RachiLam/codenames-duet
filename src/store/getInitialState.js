import { utils } from '../utils';

const makeCardState = (id) => ({
  id,
  word: utils.getWord(),
  status: [null, null],
});

const makePlayerState = (id) => ({
  id,
  successfulTurnCount: 0,
});

export const getInitialState = () => ({
  cards: utils.range(25).map((index) => makeCardState(index)),
  players: utils.range(2).map((index) => makePlayerState(index)),
  sideboard: {
    maxTurns: 9,
  },
  flags: {
    adminBoard: false,
  },
  adminMode: false,
});
