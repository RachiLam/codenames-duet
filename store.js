import Vue from 'vue';
import Vuex from 'vuex';
import dictionary from './dictionary';

const range = (count) => Array.from({ length: count }).map((value, index) => index);

const getWord = () => {
  const wordIndex = Math.floor(Math.random() * dictionary.length);
  return dictionary[wordIndex];
};

const makeCardState = (id) => ({
  id,
  word: getWord().toUpperCase(),
  isAssassin: [false, false],
  isAgent: [false, false],
  isBystander: [false, false],
});

const makePlayerState = (id) => ({
  id,
  successfulTurnCount: 0,
});

export default new Vuex.Store({
  state: {
    cards: range(25).map((index) => makeCardState(index)),
    players: range(2).map((index) => makePlayerState(index)),
    sideboard: {
      maxTurns: 9,
    },
  },
  getters: {
    turnsTaken: (state) => {
      const successfulTurns = state.players.reduce(
        (turnsTaken, player) => turnsTaken + player.successfulTurnCount,
        0,
      );

      const failedTurns = state.cards.reduce(
        (turnsTaken, card) => {
          const player0TurnsTaken = card.isBystander[0] ? 1 : 0;
          const player1TurnsTaken = card.isBystander[1] ? 1 : 0;

          return turnsTaken + player0TurnsTaken + player1TurnsTaken;
        },
        0,
      );

      return successfulTurns + failedTurns;
    },
  },
  mutations: {
    /* eslint-disable no-param-reassign */
    updateSuccessfulTurnCount(state, { playerId, count }) {
      state.players[playerId].successfulTurnCount = count;
    },
    toggleCardState(state, { cardId, key, playerId }) {
      Vue.set(state.cards[cardId][key], playerId, !state.cards[cardId][key][playerId]);
    },
    /* eslint-enable no-param-reassign */
  },
});
