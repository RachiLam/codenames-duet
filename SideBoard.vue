<template>
  <div class="player-board">
    <reset-button
      class="reset-button-top"
      :activated="resetState[0]"
      @reset="reset(0)"
    />
    <div class="turn-counters">
      <turn-counter
        v-for="index in remainingCount"
        :key="`remaining-turn-counter-${index}`"
        :type="isInErrorState ? 'error' : 'remaining'"
      />
    </div>
    <reset-button
      class="reset-button-bottom"
      :activated="resetState[1]"
      @reset="reset(1)"
    />
  </div>
</template>

<script>
import Vue from 'vue';
import ResetButton from './ResetButton.vue';
import TurnCounter from './TurnCounter.vue';

export default {
  components: {
    ResetButton,
    TurnCounter,
  },
  props: {
    sideboard: {
      type: Object,
      required: true,
    },
    turnsTaken: {
      type: Number,
      required: true,
    },
  },
  data() {
    return {
      resetState: [false, false],
    };
  },
  computed: {
    isInErrorState() {
      return this.turnsTaken > this.sideboard.maxTurns;
    },
    classes() {
      return {
        error: this.isInErrorState,
      };
    },
    remainingCount() {
      return this.isInErrorState
        ? this.turnsTaken - this.sideboard.maxTurns
        : this.sideboard.maxTurns - this.turnsTaken;
    },
  },
  methods: {
    reset(playerId) {
      Vue.set(this.resetState, playerId, !this.resetState[playerId]);

      if (this.resetState.every((value) => value)) {
        window.location.reload();
      }
    },
  },
};
</script>

<style lang="scss" scoped>
  @import 'variables.scss';

  #side-board {
    display: grid;
    grid-template-columns: 1;
    grid-template-rows: $reset-button-size auto $reset-button-size;

    .reset-button-top {
      grid-row-start: 1;
      grid-row-end: 2;
    }

    .turn-counters {
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;
      justify-content: center;
      grid-row-start: 2;
      grid-row-end: 3;

      .turn-counter {
        margin-bottom: $padding-sm;
      }
    }

    .reset-button-bottom {
      grid-row-start: 3;
      grid-row-end: 4;
    }
  }
</style>
