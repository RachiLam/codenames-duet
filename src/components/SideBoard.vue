<template>
  <div class="player-board">
    <lil-button
      class="lil-button-top"
      :activated="resetState[0]"
      @click="reset(0)"
    />
    <div class="turn-counters">
      <turn-counter
        v-for="index in remainingCount"
        :key="`remaining-turn-counter-${index}`"
        :type="isInErrorState ? 'error' : 'remaining'"
      />
    </div>
    <lil-button
      class="lil-button-bottom"
      :activated="resetState[1]"
      @click="reset(1)"
    />
  </div>
</template>

<script>
import Vue from 'vue';
import { mapGetters, mapState } from 'vuex';
import LilButton from './LilButton.vue';
import TurnCounter from './TurnCounter.vue';

export default {
  components: {
    LilButton,
    TurnCounter,
  },
  data() {
    return {
      resetState: [false, false],
    };
  },
  computed: {
    ...mapState(['sideboard']),
    ...mapGetters(['turnsTaken']),
    isInErrorState() {
      return this.turnsTaken > this.sideboard.maxTurns;
    },
    classes() {
      return {
        error: this.isInErrorState,
      };
    },
    remainingCount() {
      return this.isInErrorState
        ? this.turnsTaken - this.sideboard.maxTurns
        : this.sideboard.maxTurns - this.turnsTaken;
    },
  },
  methods: {
    reset(playerId) {
      Vue.set(this.resetState, playerId, !this.resetState[playerId]);

      if (this.resetState.every((value) => value)) {
        window.location.reload();
      }
    },
  },
};
</script>

<style lang="scss" scoped>
  @use 'styles/padding';
  @use 'styles/constants';

  #side-board {
    display: grid;
    grid-template-columns: 1;
    grid-template-rows: constants.$lil-button-size auto constants.$lil-button-size;

    .lil-button-top {
      grid-row-start: 1;
      grid-row-end: 2;
    }

    .turn-counters {
      display: flex;
      flex-direction: column;
      width: 100%;
      height: 100%;
      justify-content: center;
      grid-row-start: 2;
      grid-row-end: 3;

      .turn-counter {
        margin-bottom: padding.$sm;
      }
    }

    .lil-button-bottom {
      grid-row-start: 3;
      grid-row-end: 4;
    }
  }
</style>
