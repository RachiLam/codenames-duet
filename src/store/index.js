import Vuex from 'vuex';
import { getInitialState } from './getInitialState';
import { getters } from './getters';
import { mutations } from './mutations';
import { actions } from './actions';

export default new Vuex.Store({
  state: getInitialState(),
  getters,
  mutations,
  actions,
});
