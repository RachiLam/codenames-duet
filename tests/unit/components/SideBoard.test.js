import { getInitialState } from 'store/getInitialState';
import SideBoard from 'components/SideBoard.vue';
import ResetButton from 'components/ResetButton.vue';
import TurnCounter from 'components/TurnCounter.vue';

const { mount, shallowMount } = getMountUtils({
  component: SideBoard,
});

describe('components/SideBoard', () => {
  describe('at the beginning of the game', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount({
        store: {
          ...getInitialState(),
          turnsTaken: 0,
        },
      });
    });

    it('renders two reset buttons', () => {
      expect(wrapper.findAllComponents(ResetButton).length).toBe(2);
    });

    it('renders nine turn counters', () => {
      expect(wrapper.findAllComponents(TurnCounter).length).toBe(9);
    });

    it('renders normal counters', () => {
      const turnCounterElements = wrapper.findAllComponents(TurnCounter).wrappers
        .map((innerWrapper) => innerWrapper.element);
      expect(turnCounterElements).toMatchSnapshot();
    });
  });

  describe('when turns have been used', () => {
    let wrapper;
    beforeAll(() => {
      wrapper = shallowMount({
        store: {
          sideboard: { maxTurns: 9 },
          turnsTaken: 3,
        },
      });
    });

    it('renders less turn counters', () => {
      expect(wrapper.findAllComponents(TurnCounter).length).toBe(6);
    });
  });

  describe('when too many turns have been used', () => {
    let wrapper;
    let turnCounterElements;
    beforeAll(() => {
      wrapper = shallowMount({
        store: {
          sideboard: { maxTurns: 9 },
          turnsTaken: 12,
        },
      });

      turnCounterElements = wrapper.findAllComponents(TurnCounter).wrappers
        .map((innerWrapper) => innerWrapper.element);
    });

    it('renders the overused counters', () => {
      expect(turnCounterElements.length).toBe(3);
    });

    it('renders error counters', () => {
      expect(turnCounterElements).toMatchSnapshot();
    });
  });

  describe('when both reset buttons are pressed', () => {
    let wrapper;
    let topResetButton;
    let bottomResetButton;
    const reload = jest.fn();
    beforeAll(() => {
      wrapper = mount({
        store: {
          reload,
        },
      });

      [topResetButton, bottomResetButton] = wrapper.findAllComponents(ResetButton).wrappers;
      topResetButton.trigger('click');
      bottomResetButton.trigger('click');
    });

    it('reloads the window', () => {
      expect(reload).toHaveBeenCalled();
    });
  });

  describe('when a reset button is pressed twice before the other reset button is pressed', () => {
    let wrapper;
    let topResetButton;
    let bottomResetButton;
    const reload = jest.fn();
    beforeAll(() => {
      wrapper = mount({
        store: {
          reload,
        },
      });

      [topResetButton, bottomResetButton] = wrapper.findAllComponents(ResetButton).wrappers;
      topResetButton.trigger('click');
      topResetButton.trigger('click');
      bottomResetButton.trigger('click');
    });

    it('does not reload the window', () => {
      expect(reload).not.toHaveBeenCalled();
    });
  });
});
