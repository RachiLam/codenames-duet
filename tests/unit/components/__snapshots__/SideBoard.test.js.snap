// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/SideBoard at the beginning of the game renders normal counters 1`] = `
Array [
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
  <turn-counter-stub
    type="normal"
  />,
]
`;

exports[`components/SideBoard when too many turns have been used renders error counters 1`] = `
Array [
  <turn-counter-stub
    type="error"
  />,
  <turn-counter-stub
    type="error"
  />,
  <turn-counter-stub
    type="error"
  />,
]
`;
